merged with trunk to fix the failure. 

one test failure against s3 london, ` -Dparallel-tests -DtestsThreadCount=10`

```
[INFO] Running org.apache.hadoop.fs.s3a.commit.ITestCommitOperationCost
[ERROR] Tests run: 44, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 79.159 s <<< FAILURE! - in org.apache.hadoop.fs.contract.s3a.ITestS3AContractVectoredRead
[ERROR] testStopVectoredIoOperationsUnbuffer[Buffer type : direct](org.apache.hadoop.fs.contract.s3a.ITestS3AContractVectoredRead)  Time elapsed: 1.26 s  <<< FAILURE!
java.lang.AssertionError: Expected an exception of type class java.io.InterruptedIOException
        at org.apache.hadoop.test.LambdaTestUtils.intercept(LambdaTestUtils.java:409)
        at org.apache.hadoop.fs.contract.s3a.ITestS3AContractVectoredRead.testStopVectoredIoOperationsUnbuffer(ITestS3AContractVectoredRead.java:143)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.lang.Thread.run(Thread.java:750)
```
anyone else seen this? race condition maybe?

--- a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestWeakReferenceMap.java
+++ b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestWeakReferenceMap.java
@@ -21,7 +21,6 @@ package org.apache.hadoop.util;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicLong;
-import java.util.stream.IntStream;
 
 import org.assertj.core.api.Assertions;
 import org.junit.Before;

--- a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestWeakReferenceMap.java
+++ b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestWeakReferenceMap.java
@@ -20,6 +20,8 @@ package org.apache.hadoop.util;
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.IntStream;
 
 import org.assertj.core.api.Assertions;
 import org.junit.Before;
@@ -144,11 +146,16 @@ public class TestWeakReferenceMap extends AbstractHadoopTestBase {
    * Test the WeakReferenceThreadMap extension.
    */
   @Test
-  public void testWeakReferenceThreadMapRejectsNullAssignment()
+  public void testWeakReferenceThreadMapAssignment()
       throws Throwable {
+
+    // counters foor the callbacks
+    final AtomicLong created = new AtomicLong();
+    final AtomicLong lost = new AtomicLong();
+
     WeakReferenceThreadMap<String> threadMap = new WeakReferenceThreadMap<>(
-        id -> "Entry for thread ID " + id,
-        null);
+        id -> "Entry for thread ID " + id + " (" + created.incrementAndGet() + ")",
+        id -> lost.incrementAndGet());
 
     Assertions.assertThat(threadMap.setForCurrentThread("hello"))
         .describedAs("current thread map value on first set")
@@ -179,15 +186,29 @@ public class TestWeakReferenceMap extends AbstractHadoopTestBase {
         .isNull();
 
     // lookup will return a new instance created by the factory
+    long c1 = created.get();
     String dynamicValue = threadMap.getForCurrentThread();
     Assertions.assertThat(dynamicValue)
         .describedAs("dynamically created thread map value")
-        .startsWith("Entry for thread ID");
+        .startsWith("Entry for thread ID")
+        .contains("(" + (c1 + 1) + ")");
 
     // and we can overwrite that
     Assertions.assertThat(threadMap.setForCurrentThread("hello2"))
         .describedAs("value before the thread entry is changed")
         .isEqualTo(dynamicValue);
+
+    // simulate a weak gc
+    long threadId = threadMap.currentThreadId();
+    threadMap.put(threadId, null);
+    String updated = threadMap.getForCurrentThread();
+    Assertions.assertThat(lost.get())
+        .describedAs("lost count")
+        .isEqualTo(1);
+    Assertions.assertThat(updated)
+        .describedAs("dynamically created thread map value")
+        .startsWith("Entry for thread ID")
+        .contains("(" + (c1 + 2) + ")");
   }
 
   /**

